### PR TITLE
Streamline data provider information reading

### LINKED ISSUE #1691: Defining dataProviderClass in a subclass
I'm not sure if this is even possible, but I've got a few test cases where I can run the same tests for two implementations of a class, however they require different input data. So I've created a shared superclass for both tests, and set the tests with ``@Test(dataProvider = "myProvider")``. Then, in the actual test classes, I set ``@Test(dataProviderClass = Provider.class)``. Provider has a method annotated as a ``@DataProvider`` and named ``myProvider``
### TestNG Version
6.14.2
### Expected behavior
TestNG should pick up the dataProviderClass set in the subclasses

### Actual behavior
TestNG fails with an error (stacktrace is attached in the gist below), complaining about a missing dataprovider.

Things work fine if I move the methods from ``Provider`` into the test classes as well, it just doesn't seem to work when I separate the providers into a separate class.

### Is the issue reproductible on runner?

- [ ] Shell
- [x] Maven
- [ ] Gradle
- [ ] Ant
- [ ] Eclipse
- [x] IntelliJ
- [ ] NetBeans

### Test case sample
https://gist.github.com/sonOfRa/dd337372e75b8a86fa68aeac6d0094ef

### PR DESCRIPTION
**Pull request by krmahadevan**
Closes #1691

Fixes #1691  .

### Did you remember to?

- [X] Add test case(s)
- [X] Update `CHANGES.txt`

We encourage pull requests that:

* Add new features to TestNG (or)
* Fix bugs in TestNG

If your pull request involves fixing SonarQube issues then we would suggest that you please discuss this with the 
[TestNG-dev](https://groups.google.com/forum/#!forum/testng-dev) before you spend time working on it.

### CODE REVIEW

**Comment by krmahadevan** [comment]
ping @juherr

**Review by juherr — CHANGES_REQUESTED** [review]
Please, add more test cases. 
For example, check the behavior when both class and method have a configured `dataProviderClass` and/or a `dataProvider`.

File: src/main/java/org/testng/internal/Parameters.java [hunk 1]
```diff
@@ -591,32 +591,55 @@ private static IDataProviderMethod findDataProvider(
...
+      methodLevel.setDataProvider(classLevel.getDataProvider());
+    }
+    if (isDataProviderClassEmpty(methodLevel) && !isDataProviderClassEmpty(classLevel)) {
+      methodLevel.setDataProviderClass(classLevel.getDataProviderClass());
+    }
+    return methodLevel;
+  }
+
+  private static boolean isDataProviderClassEmpty(ITestAnnotation annotation) {
+    return annotation.getDataProviderClass() == null || Object.class.equals(annotation.getDataProviderClass());
```
[the remark this answers is not in the document]
Comment (krmahadevan) [reply src/main/java/org/testng/internal/Parameters.java]: We can't use `Void` because our annotation by default sets the value to `Object`.

**Comment by krmahadevan** [comment]
@juherr - I have added more tests. Please take a look.

File: src/main/java/org/testng/internal/Parameters.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (juherr) [reply src/main/java/org/testng/internal/Parameters.java]: The idea was more to already support Void now and remove Object after.

**Review by juherr — COMMENTED** [review]
What is the behavior if a class has configured a data provider but test methods don't need it?

The expected behavior is running the tests without any problem.
We can add a test to confirm the behavior.

I wonder if using a data provider on class and having no param method or overring a data provider is a really good idea. Maybe we should warn users against it.

**Comment by krmahadevan** [comment]
@juherr 

> What is the behavior if a class has configured a data provider but test methods don't need it?

That was a good catch. I wrote a test for this. 

> The expected behavior is running the tests without any problem.

I dont think that can be made achievable. When a test method inherits a data provider from a class level, and if it doesn't have any arguments defined on it and if the data provider ends up sending argument values, which the test method can't receive, we will now throw an error.  The error message will be verbose enough to let the user know what the issue was. 

> We can add a test to confirm the behavior.

yes I have added a test to confirm the new behavior (that is, the test method errors out) and not the behavior which you are suggesting (which is the test should run without any problem)

I have pushed in the changes as an additional commit. Please help take a look

**Review by juherr — APPROVED** [review]

**Comment by juherr** [comment]
We should just add a warning in the release note because it will break backward compatibility because current tests may have a data provider on class without problem and it will fail with the next release.

**Comment by krmahadevan** [comment]
> We should just add a warning in the release note because it will break backward compatibility because current tests may have a data provider on class without problem and it will fail with the next release.

Sure. I will work on starting to compile a release notes which talks about the changes for `7.0.0` and include a reference to this as well.

**Pull request merged**